### PR TITLE
Add Prometheus Remote Write Exporter supporting Cortex - conversion and export for Histogram OTLP metrics

### PR DESCRIPTION
**Pull request by huyan0**
This PR is part of a series of PRs implementing a Prometheus remote write exporter supporting Cortex. 
**See  related PR #1577**  

**Description:** This PR adds export support for histogram and summary metrics for Prometheus remote write [integrated backends](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage), such as Cortex. The exporter drops non-cumulative monotonic, histogram, and summary OTLP metrics.

Please note this metrics exporter does not support Prometheus default labels such as `job` or `instance` labels. An issue addressing Prometheus default labels will be filed later. Another related feature is to derive labels from a Resource. This functionality already exists in the Go SDK and will be implemented in another PR. This feature could allow users to specify which attributes they want to add as labels.

**Link to tracking Issue:**  #1150
Related issues are: 
Metrics aggregation proposal: #1422
Prometheus exporter not functional: #1255
Related spec discussion: [#731](https://github.com/open-telemetry/opentelemetry-specification/issues/731)

**Documentation:** 

* readme.md with sample configuration, file structure, and assumptions

* Design documentation: PR #1464

cc: @huyan0 @alolita @jmacd @bogdandrutu

### CODE REVIEW

**Comment by bogdandrutu** [comment]
Please split this into 2 PRs the first being the Histogram. We will not support Summary for few weeks in the new otlp and that will get delayed.

**Comment by bogdandrutu** [comment]
Please increase the coverage :)

**Comment by huyan0** [comment]
> Please increase the coverage :)

Will do. Also rebasing : )

**Comment by huyan0** [comment]
Got 100% diff coverage now

File: exporter/prometheusremotewriteexporter/exporter.go [hunk 1]
```diff
@@ -19,46 +19,47 @@ import (
...
 
 	if client == nil {
-		return nil, errors.New("http client cannot be nil")
+		return nil, fmt.Errorf("http client cannot be nil")
 	}
 
 	endpointURL, err := url.ParseRequestURI(endpoint)
 	if err != nil {
-		return nil, errors.New("invalid endpoint")
+		return nil, fmt.Errorf("invalid endpoint")
```
Comment (bogdandrutu) [inline exporter/prometheusremotewriteexporter/exporter.go:59]: Use `errors.New` when no need for printf like. And make sure you use the new errors package from golang

File: exporter/prometheusremotewriteexporter/exporter.go [hunk 2]
```diff
@@ -138,19 +148,21 @@ func (prwe *prwExporter) pushMetrics(ctx context.Context, md pdata.Metrics) (int
...
 	// int points
 	case otlp.MetricDescriptor_MONOTONIC_INT64, otlp.MetricDescriptor_INT64:
 		if metric.Int64DataPoints == nil {
-			return errors.New("nil data point field in metric" + metric.GetMetricDescriptor().Name)
+			return fmt.Errorf("nil data point field in metric %v", metric.GetMetricDescriptor().Name)
 		}
 
 		for _, pt := range metric.Int64DataPoints {
-
+			if pt == nil {
```
Comment (bogdandrutu) [inline exporter/prometheusremotewriteexporter/exporter.go:163]: This is not from this change please extract in a separate PR.

File: exporter/prometheusremotewriteexporter/exporter.go [hunk 3]
```diff
@@ -19,46 +19,47 @@ import (
...
 	"go.opentelemetry.io/collector/consumer/pdata"
 	"go.opentelemetry.io/collector/consumer/pdatautil"
 	otlp "go.opentelemetry.io/collector/internal/data/opentelemetry-proto-gen/metrics/v1old"
 	"go.opentelemetry.io/collector/internal/dataold"
 )
 
-// prwExporter converts OTLP metrics to Prometheus remote write TimeSeries and sends them to a remote endpoint
-type prwExporter struct {
+// PrwExporter converts OTLP metrics to Prometheus remote write TimeSeries and sends them to a remote endpoint
+type PrwExporter struct {
```
[the remark this answers is not in the document]
Comment (huyan0) [reply exporter/prometheusremotewriteexporter/exporter.go:41]: We are trying to import this exporter in another project and it would be really helpful if we can import this dependency here. This is for the authentication requirement we have discussed about.

File: exporter/prometheusremotewriteexporter/exporter.go [hunk 4]
```diff
@@ -19,46 +19,47 @@ import (
...
 	client      *http.Client
 	wg          *sync.WaitGroup
 	closeChan   chan struct{}
 }
 
-// newPrwExporter initializes a new prwExporter instance and sets fields accordingly.
+// NewPrwExporter initializes a new PrwExporter instance and sets fields accordingly.
 // client parameter cannot be nil.
-func newPrwExporter(namespace string, endpoint string, client *http.Client) (*prwExporter, error) {
+func NewPrwExporter(namespace string, endpoint string, client *http.Client) (*PrwExporter, error) {
```
[the remark this answers is not in the document]
Comment (huyan0) [reply exporter/prometheusremotewriteexporter/exporter.go:51]: same as the last comment

File: exporter/prometheusremotewriteexporter/exporter.go (hunk 2 above)
Comment (huyan0) [reply exporter/prometheusremotewriteexporter/exporter.go:163]: Sure, will do.

File: exporter/prometheusremotewriteexporter/exporter.go [hunk 5]
```diff
@@ -4,7 +4,7 @@
 // you may not use this file except in compliance with the License.
 // You may obtain a copy of the License at
 //
-//      http://www.apache.org/licenses/LICENSE-2.0
+//       http://www.apache.org/licenses/LICENSE-2.0
```
[the remark this answers is not in the document]
Comment (huyan0) [reply exporter/prometheusremotewriteexporter/exporter.go:7]: Mostly formatting changes, will remove.

**Comment by huyan0** [comment]
Addressed some comments. Really appreciate the review. Will start working on #1681

File: exporter/prometheusremotewriteexporter/exporter.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply exporter/prometheusremotewriteexporter/exporter.go:42]: Sure, that may be fine but do a separate PR please

[the remark this answers is not in the document]
Comment (huyan0) [reply exporter/prometheusremotewriteexporter/exporter.go]: Cool. I have changed them back here. Thanks.

File: exporter/prometheusremotewriteexporter/exporter.go [hunk 6]
```diff
@@ -146,11 +157,10 @@ func (prwe *prwExporter) handleScalarMetric(tsMap map[string]*prompb.TimeSeries,
 	// int points
 	case otlp.MetricDescriptor_MONOTONIC_INT64, otlp.MetricDescriptor_INT64:
 		if metric.Int64DataPoints == nil {
-			return errors.New("nil data point field in metric" + metric.GetMetricDescriptor().Name)
+			return fmt.Errorf("nil data point field in metric %v", metric.GetMetricDescriptor().Name)
```
Comment (bogdandrutu) [inline exporter/prometheusremotewriteexporter/exporter.go:160]: %s?

File: exporter/prometheusremotewriteexporter/testutil_test.go [hunk 7]
```diff
@@ -156,6 +156,43 @@ func getDoubleDataPoint(labels []*commonpb.StringKeyValue, value float64, ts uin
...
+		StartTimeUnixNano: 0,
+		TimeUnixNano:      ts,
+		Count:             count,
+		Sum:               sum,
+		Buckets:           bks,
+		ExplicitBounds:    bounds,
+	}
+}
+
+func getSummaryDataPoint(labels []*commonpb.StringKeyValue, ts uint64, sum float64, count uint64, pcts []float64, values []float64) *otlp.SummaryDataPoint {
```
Comment (bogdandrutu) [inline exporter/prometheusremotewriteexporter/testutil_test.go:178]: Unnecessary?

Comment (huyan0) [reply exporter/prometheusremotewriteexporter/testutil_test.go:178]: This is used for the test case where the metric type is not supported.

Comment (huyan0) [reply exporter/prometheusremotewriteexporter/testutil_test.go]: This case is still tested. Instead of initializing a point I just put in an empty summary.

**Comment by vishiy** [comment]
@huyan0 - I see the prometheus histograms represented as OTEL histograms, during which the values are becoming delta ( see design here - https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/prometheusreceiver/DESIGN.md#histogram ). I see the same issue with summary (for _sum & _count) and also counters. Is there a way to configure the receiver to NOT do the delta/diff from first scrape ? If not, how to get the original values as scraped for these metrics ?
